flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java

```diff
@@ -0,0 +1,323 @@
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.SHUFFLE_MEMORY_MIN.key())), is(TM_RESOURCE_SPEC.getShuffleMemSize()));
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.MANAGED_MEMORY_SIZE.key())), is(TM_RESOURCE_SPEC.getManagedMemorySize()));
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.MANAGED_MEMORY_OFFHEAP_SIZE.key())), is(TM_RESOURCE_SPEC.getOffHeapManagedMemorySize()));
+	}
+
+	@Test
+	public void testGenerateJvmParameters() throws Exception {
+		String jvmParamsStr = TaskExecutorResourceUtils.generateJvmParametersStr(TM_RESOURCE_SPEC);
+		MemorySize heapSizeMax = null;
+		MemorySize heapSizeMin = null;
+		MemorySize directSize = null;
+		MemorySize metaspaceSize = null;
+		for (String paramStr : jvmParamsStr.split(" ")) {
+			if (paramStr.startsWith("-Xmx")) {
+				heapSizeMax = MemorySize.parse(paramStr.substring("-Xmx".length()));
+			} else if (paramStr.startsWith("-Xms")) {
+				heapSizeMin = MemorySize.parse(paramStr.substring("-Xms".length()));
+			} else if (paramStr.startsWith("-XX:MaxDirectMemorySize=")) {
+				directSize = MemorySize.parse(paramStr.substring("-XX:MaxDirectMemorySize=".length()));
+			} else if (paramStr.startsWith("-XX:MetaspaceSize=")) {
+				metaspaceSize = MemorySize.parse(paramStr.substring("-XX:MetaspaceSize=".length()));
+			} else {
+				throw new Exception("Unknown JVM parameter: " + paramStr);
+			}
+		}
+
+		assertThat(heapSizeMax, is(TM_RESOURCE_SPEC.getFrameworkHeapSize().add(TM_RESOURCE_SPEC.getTaskHeapSize()).add(TM_RESOURCE_SPEC.getOnHeapManagedMemorySize())));
+		assertThat(heapSizeMin, is(heapSizeMax));
+		assertThat(directSize, is(TM_RESOURCE_SPEC.getTaskOffHeapSize().add(TM_RESOURCE_SPEC.getShuffleMemSize())));
+		assertThat(metaspaceSize, is(TM_RESOURCE_SPEC.getJvmMetaspaceSize()));
+	}
+
+	@Test
+	public void testConfigFrameworkHeapMemory() throws Exception {
+		final MemorySize frameworkHeapSize = MemorySize.parse("100m");
+
+		Configuration conf = new Configuration();
+		conf.setString(TaskManagerOptions.FRAMEWORK_HEAP_MEMORY, frameworkHeapSize.getMebiBytes() + "m");
+
+		validateInAllConfigurations(conf, taskExecutorResourceSpec -> assertThat(taskExecutorResourceSpec.getFrameworkHeapSize(), is(frameworkHeapSize)));
+	}
+
+	@Test
+	public void testConfigTaskHeapMemory() throws Exception {
+		final MemorySize taskHeapSize = MemorySize.parse("50m");
+
+		Configuration conf = new Configuration();
+		conf.setString(TaskManagerOptions.TASK_HEAP_MEMORY, taskHeapSize.getMebiBytes() + "m");
+
+		// validate only in configuration with explicit task heap and managed memory size,
+		// in other cases, heap memory is derived from total flink memory, thus the configured value is not used by design
+		validateInConfigWithExplicitTaskHeapAndManagedMem(conf, taskExecutorResourceSpec -> assertThat(taskExecutorResourceSpec.getTaskHeapSize(), is(taskHeapSize)));
```

Comment:
I find it confusing that certain configuration values are only valid in certain scenarios.